psi4/src/psi4/libmints/mintshelper.cc

```diff
@@ -1996,4 +2046,1146 @@ void MintsHelper::play()
+                        grad[0]->set(p+oP, q+oQ,(*ref0++));
+                        grad[1]->set(p+oP, q+oQ,(*ref1++));
+                        grad[2]->set(p+oP, q+oQ,(*ref2++));
+                }
+        }
+
+        return grad;
+}
+
+
+std::vector<SharedMatrix> MintsHelper::ao_potential_deriv2_helper(int atom1, int atom2)
+{
+
+    char lbl[32];
+    char ** cartcomp;
+    cartcomp = (char **) malloc (3 * sizeof(char *));
+    cartcomp[0] = strdup("x");
+    cartcomp[1] = strdup("y");
+    cartcomp[2] = strdup("z");
+  
+    std::shared_ptr<OneBodyAOInt> Vint(integral_->ao_potential(2));
+
+    std::shared_ptr <BasisSet> bs1 = Vint->basis1();
+    std::shared_ptr <BasisSet> bs2 = Vint->basis2();
+
+    int nbf1 = bs1->nbf();
+    int nbf2 = bs2->nbf();
+    int natom = molecule_->natom();
+
+    std::vector<SharedMatrix> grad;
+    for (int a=0,ab=0; a<3; a++)
+      for (int b=0; b<3; b++,ab++){
+        sprintf(lbl, "ao_potential_deriv2_%d_%d_%s_%s", atom1, atom2, cartcomp[a], cartcomp[b]);
+        grad.push_back(SharedMatrix(new Matrix(lbl, nbf1, nbf2)));
+        grad[ab]->zero();    
+      }
+
+    const double *buffer = Vint->buffer();
+
+    for (int P = 0; P < bs1->nshell(); P++) {
+            for (int Q = 0; Q < bs2->nshell(); Q++) {
+
+                int nP = bs1->shell(P).nfunction();
+                int oP = bs1->shell(P).function_index();
+                int aP = bs1->shell(P).ncenter();
+
+                int nQ = bs2->shell(Q).nfunction();
+                int oQ = bs2->shell(Q).function_index();
+                int aQ = bs2->shell(Q).ncenter();
+
+                size_t offset = nP*nQ;
+
+                std::map<std::string, double> grad_map;
+                for (int a=0; a<3; a++)
+                    for (int b=0; b<3; b++){
+                        std::string grad_key = std::string(cartcomp[a]) + std::string(cartcomp[b]);
+                        grad_map[grad_key] = 0;
+                    }
+
+                for(int atom = 0; atom < natom; atom++){
+
+                   double Z = molecule_->Z(atom);
+                   Vint->set_origin(molecule_->xyz(atom));
+
+                   std::pair <int, int> target_atoms(atom1, atom2);  
+                   std::vector<std::pair<int, int> > vec_pairs; 
+                   std::vector<int> vec_pos; 
+                   std::map<int, std::string> pos_map;
+                   std::string strings;
+                   std::string key; 
+
+                   vec_pairs.push_back(std::make_pair(aP,aP)); 
+                   vec_pairs.push_back(std::make_pair(aQ,aQ)); 
+                   vec_pairs.push_back(std::make_pair(atom,atom)); 
+                   vec_pairs.push_back(std::make_pair(atom,aP)); 
+                   vec_pairs.push_back(std::make_pair(aP,aQ)); 
+                   vec_pairs.push_back(std::make_pair(atom,aQ)); 
+
+
+                   for (std::vector<std::pair<int, int> >::iterator it = vec_pairs.begin() ; it != vec_pairs.end(); ++it){
```

Comment:
Could be nice to use a ranged based for loop here.

Reply:
I need the position of the elements of the container here. So a range based for loop may not be ideal here.